tls_manager.go

```diff
@@ -0,0 +1,517 @@
+package lnd
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/lightningnetwork/lnd/cert"
+	"github.com/lightningnetwork/lnd/keychain"
+	"github.com/lightningnetwork/lnd/lncfg"
+	"github.com/lightningnetwork/lnd/lnencrypt"
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"golang.org/x/crypto/acme/autocert"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+const (
+	// modifyFilePermissons is the file permission used for writing
+	// encrypted tls files.
+	modifyFilePermissions = 0600
+
+	// validityHours is the number of hours the ephemeral tls certificate
+	// will be valid, if encrypting tls certificates is turned on.
+	validityHours = 24
+)
+
+var (
+	// privateKeyPrefix is the prefix to a plaintext TLS key.
+	privateKeyPrefix = []byte("-----BEGIN EC PRIVATE KEY-----")
+)
+
+// TLSManager generates/renews a TLS cert/key pair when needed. When required,
+// it encrypts the TLS key. It also returns the certificate configuration
+// options needed for gRPC and REST.
+type TLSManager struct {
+	cfg         *Config
+	tlsReloader *cert.TLSReloader
```

Comment:
Would be great if you could document `tlsReloader` as switching mechanics of certificates is still not 100% clear to me.